File: app/src/main/java/ml/melun/mangaview/download/DownloadQueueWorker.java
package ml.melun.mangaview.download;

import android.content.Context;
import android.content.Intent;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import java.io.File;
import java.io.FileInputStream;
import java.io.ByteArrayOutputStream;

import ml.melun.mangaview.Downloader;
import ml.melun.mangaview.report.CrashReporter;

public class DownloadQueueWorker extends Worker {
    public static final String KEY_QUEUE_ID = "queue_id";
    public static final String QUEUE_DIR = "download_queue";

    public DownloadQueueWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        String queueId = getInputData().getString(KEY_QUEUE_ID);
        if(queueId == null || queueId.length() == 0)
            return Result.failure();
        File file = new File(getApplicationContext().getFilesDir(), QUEUE_DIR + "/" + queueId + ".json");
        if(!file.exists())
            return Result.failure();
        try {
            String payload = readPayload(file);
            int split = payload.indexOf('\n');
            if(split <= 0)
                return Result.failure();
            Intent intent = new Intent(getApplicationContext(), Downloader.class);
            intent.setAction(Downloader.ACTION_QUEUE);
            intent.putExtra("title", payload.substring(0, split));
            intent.putExtra("selected", payload.substring(split + 1));
            if(Build.VERSION.SDK_INT >= 26)
                getApplicationContext().startForegroundService(intent);
            else
                getApplicationContext().startService(intent);
            file.delete();
            return Result.success();
        } catch (Exception e) {
            CrashReporter.record(e);
            return Result.retry();
        }
    }

    private String readPayload(File file) throws java.io.IOException {
        try(FileInputStream input = new FileInputStream(file);
            ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[8192];
            int read;
            while((read = input.read(buffer)) > 0)
                output.write(buffer, 0, read);
            return output.toString();
        }
    }
}

File: app/src/main/java/ml/melun/mangaview/download/DownloadQueueWorker.kt
package ml.melun.mangaview.download

import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.work.Worker
import androidx.work.WorkerParameters
import ml.melun.mangaview.Downloader
import java.io.File

class DownloadQueueWorker(
    appContext: Context,
    params: WorkerParameters,
) : Worker(appContext, params) {
    override fun doWork(): Result {
        val queueId = inputData.getString(KEY_QUEUE_ID) ?: return Result.failure()
        val file = File(applicationContext.filesDir, "$QUEUE_DIR/$queueId.json")
        if (!file.exists()) return Result.failure()
        return try {
            val payload = file.readText()
            val split = payload.indexOf('\n')
            if (split <= 0) return Result.failure()
            val title = payload.substring(0, split)
            val selected = payload.substring(split + 1)
            val intent = Intent(applicationContext, Downloader::class.java).apply {
                action = Downloader.ACTION_QUEUE
                putExtra("title", title)
                putExtra("selected", selected)
            }
            if (Build.VERSION.SDK_INT >= 26) {
                applicationContext.startForegroundService(intent)
            } else {
                applicationContext.startService(intent)
            }
            file.delete()
            Result.success()
        } catch (_: Exception) {
            Result.retry()
        }
    }

    companion object {
        const val KEY_QUEUE_ID = "queue_id"
        const val QUEUE_DIR = "download_queue"
    }
}
